editoast/src/main.rs

```diff
@@ -118,6 +124,56 @@ fn init_sentry(args: &RunserverArgs) -> Option<ClientInitGuard> {
     }
 }
 
+#[derive(Debug, Serialize, Deserialize)]
+struct Claims {
+    sub: String,
+    exp: usize,
+}
+
+#[derive(Debug, Deserialize)]
+pub struct Jwk {
+    kty: String,
+    #[serde(rename = "use")]
+    jwt_use: String,
+    n: String,
+    e: String,
+    kid: String,
+    alg: Option<String>,
+}
+
+#[derive(Debug, Deserialize)]
+struct JwksResponse {
+    keys: Vec<Jwk>,
+}
+
+async fn get_jwks(domain: &str) -> Result<Vec<Jwk>, reqwest::Error> {
+    let jwks_uri = format!("{}.well-known/jwks.json", domain);
+    let jwks: JwksResponse = reqwest::get(&jwks_uri).await?.json().await?;
+    Ok(jwks.keys)
+}
+
+async fn validator(
+    req: ServiceRequest,
+    credentials: BearerAuth,
+) -> Result<ServiceRequest, (ActixError, ServiceRequest)> {
+    // let secret = env::var("OSRD_AUTH_ZERO_SECRET").expect("auth0 secret must be set");
+    let domain = env::var("OSRD_AUTH_ZERO_DOMAIN").expect("auth0 domain must be set");
+    let mut validation = Validation::new(Algorithm::RS256);
+    validation.set_issuer(&[&domain]);
+
+    let jwks = get_jwks(&domain).await.expect("Failed to fetch JWKS");
+    // let decoding_key = DecodingKey::from_secret(secret.as_ref());
+    let decoding_key = DecodingKey::from_rsa_components(&jwks[0].n, &jwks[0].e).unwrap();
+
+    match decode::<Claims>(&credentials.token(), &decoding_key, &validation) {
```

Comment:
We need to know which key to use before decoding the token.

- Use [jsonwebtoken::decode_header](https://docs.rs/jsonwebtoken/latest/jsonwebtoken/fn.decode_header.html#) to retrieve the `kid` of the token. This will retrieve the public key associated with the private key used to sign the token.
- Use the [find method](https://docs.rs/jsonwebtoken/latest/jsonwebtoken/jwk/struct.JwkSet.html#method.find) to retrieve the Jwk.
- Use the `from_jwk` method. Check documentation: https://docs.rs/jsonwebtoken/latest/jsonwebtoken/struct.DecodingKey.html#method.from_jwk. (This will automatically check the algorithm used by the public key.)
- Then call the decode function.